docs/proposal/jakarta_messaging_3_1_async_producer/jakarta_messaging_3_1_async_producer.md

```diff
@@ -0,0 +1,227 @@
+
+**2\. Approach 2: Leave it as it is and honour `isCopyMessageOnSend`:**
+
+If `CopyMessageOnSend` is false, then we don’t make a copy and set the message to be inaccessible and let the marshaller encode it (the field will be ignored anyway since we are not changing the marshaller logic).
+
+## **Messages ordering** {#messages-ordering}
+
+This corresponds to [7.3.3](https://jakarta.ee/specifications/messaging/3.1/jakarta-messaging-spec-3.1#message-order-2) and [7.3.8](https://jakarta.ee/specifications/messaging/3.1/jakarta-messaging-spec-3.1#use-of-the-completionlistener-by-the-jakarta-messaging-provider). Basically it needs to meet two cases of messaging ordering requirements:
+
+1. The order of asynchronous sends to a particular destination from a producer has to be the same as the order of execution of its corresponding CompletionListener.
+2. The order of asynchronous sends across a set of destinations from the same producer has to be the same as the order or execution of its corresponding CompletionListener.
+
+Requirement \#1 is already satisfied because the broker acknowledges messages in the same order as the messages were sent within the same destination. The client doesn’t need any additional logic to handle that.
+
+To satisfy requirement \#2, there are a few approaches.
+
+**\[Recommended\] Approach 1: Application thread send async message one at a time**
+
+This is the most straightforward option, sending async messages one at a time in the application thread. It basically makes async send behave like sync send. CompletionListener still needs to be executed on another thread (meeting the threading restriction 7.3.7), in this case the transport thread.   
+Pros:
+
+- It is very simple to implement (also makes implementing other features simpler, see the next design decision below)
+
+Cons:
+
+- It sort of defies the purpose of async send because the application thread needs to wait before the send is completed (can lead to performance issues). Even though it doesn’t violate the spec.
+
+This approach is recommended due to its simplicity. Will follow up with a plan to optimize such behaviour.
+
+**Approach 2: Application thread submits the async send to a work queue, a single background thread will dequeue a message from the queue, send it asynchronously and wait for it to finish, before dequeuing the next.**
+
+This is similar to approach 1, the async message is sent one at a time. The difference being instead of doing it in the application thread, the `Session` object will have a `SingleThreadExecutor` to handle sending of the messages on a background thread. The advantage of doing that is the application thread is not blocked by sending an async message, it simply enqueues the message to the work queue (ExecutorService) then proceeds. Allowing it to batch send a lot of async messages at a time without blocking.
```

Comment:
each send on the executor would still need a round trip to the server on this case.
you are only unblocking the sender's thread., right?

Reply:
wouldn't be possible to have a completion callback on the storage, sending responses to the client upon callback?